numpy/linalg/umath_linalg.c.src

```diff
@@ -3235,160 +3269,1005 @@ static void
+}
 
-static char eighvals_types[] = {
-    NPY_FLOAT, NPY_FLOAT,
-    NPY_DOUBLE, NPY_DOUBLE,
-    NPY_CFLOAT, NPY_FLOAT,
-    NPY_CDOUBLE, NPY_DOUBLE
-};
+/**end repeat**/
 
-static char eig_types[] = {
-    NPY_FLOAT, NPY_CFLOAT, NPY_CFLOAT,
-    NPY_DOUBLE, NPY_CDOUBLE, NPY_CDOUBLE,
-    NPY_CDOUBLE, NPY_CDOUBLE, NPY_CDOUBLE
-};
+/**begin repeat
+   #TYPE=CFLOAT,CDOUBLE#
+   #lapack_func=cgeqrf,zgeqrf#
+   #ftyp=fortran_complex,fortran_doublecomplex#
+ */
+static inline int
+init_@lapack_func@(GEQRF_PARAMS_t *params,
+                   fortran_int m,
+                   fortran_int n)
+{
+    npy_uint8 *mem_buff = NULL;
+    npy_uint8 *mem_buff2 = NULL;
+    npy_uint8 *a, *tau, *work;
+    fortran_int min_m_n = fortran_int_min(m, n);
+    size_t safe_min_m_n = min_m_n;
+    size_t safe_m = m;
+    size_t safe_n = n;
 
-static char eigvals_types[] = {
-    NPY_FLOAT, NPY_CFLOAT,
-    NPY_DOUBLE, NPY_CDOUBLE,
-    NPY_CDOUBLE, NPY_CDOUBLE
-};
+    size_t a_size = safe_m * safe_n * sizeof(@ftyp@);
+    size_t tau_size = safe_min_m_n * sizeof(@ftyp@);
 
-static char svd_1_1_types[] = {
-    NPY_FLOAT, NPY_FLOAT,
-    NPY_DOUBLE, NPY_DOUBLE,
-    NPY_CFLOAT, NPY_FLOAT,
-    NPY_CDOUBLE, NPY_DOUBLE
-};
+    fortran_int work_count;
+    size_t work_size;
+    fortran_int lda = fortran_int_max(1, m);
 
-static char svd_1_3_types[] = {
-    NPY_FLOAT,   NPY_FLOAT,   NPY_FLOAT,  NPY_FLOAT,
-    NPY_DOUBLE,  NPY_DOUBLE,  NPY_DOUBLE, NPY_DOUBLE,
-    NPY_CFLOAT,  NPY_CFLOAT,  NPY_FLOAT,  NPY_CFLOAT,
-    NPY_CDOUBLE, NPY_CDOUBLE, NPY_DOUBLE, NPY_CDOUBLE
-};
+    mem_buff = malloc(a_size + tau_size);
 
-/*  A,           b,           rcond,      x,           resid,      rank,    s,        */
-static char lstsq_types[] = {
-    NPY_FLOAT,   NPY_FLOAT,   NPY_FLOAT,  NPY_FLOAT,   NPY_FLOAT,  NPY_INT, NPY_FLOAT,
-    NPY_DOUBLE,  NPY_DOUBLE,  NPY_DOUBLE, NPY_DOUBLE,  NPY_DOUBLE, NPY_INT, NPY_DOUBLE,
-    NPY_CFLOAT,  NPY_CFLOAT,  NPY_FLOAT,  NPY_CFLOAT,  NPY_FLOAT,  NPY_INT, NPY_FLOAT,
-    NPY_CDOUBLE, NPY_CDOUBLE, NPY_DOUBLE, NPY_CDOUBLE, NPY_DOUBLE, NPY_INT, NPY_DOUBLE,
-};
+    if (!mem_buff)
+        goto error;
 
-typedef struct gufunc_descriptor_struct {
-    char *name;
-    char *signature;
-    char *doc;
-    int ntypes;
-    int nin;
-    int nout;
-    PyUFuncGenericFunction *funcs;
-    char *types;
-} GUFUNC_DESCRIPTOR_t;
+    a = mem_buff;
+    tau = a + a_size;
+    memset(tau, 0, tau_size);
+
+
+    params->M = m;
+    params->N = n;
+    params->A = a;
+    params->TAU = tau;
+    params->LDA = lda;
+
+    {
+        /* compute optimal work size */
+
+        @ftyp@ work_size_query;
+
+        params->WORK = &work_size_query;
+        params->LWORK = -1;
+
+        if (call_@lapack_func@(params) != 0)
+            goto error;
+
+        work_count = (fortran_int) ((@ftyp@*)params->WORK)->r;
+
+    }
+
+    params->LWORK = fortran_int_max(fortran_int_max(1, n),
+                                    work_count > 0 ? work_count : -work_count);
```

Comment:
What does negative work_count mean here>

Reply:
I have implemented [this line](https://github.com/numpy/numpy/pull/19151/files#diff-a5aa16fdab19a25c2c3c370e09bf02278a3d3118f5596724bffd91ec51419f65L973) in C.

Reply:
I will try removing this to see if things work without it. If so, then I will push the changes here.

Reply:
I think we may need this. Tests are failing on 32bit windows after this change. Let me push once again after re-introducing this part.

Reply:
The failed test on windows is `TestMatmul.test_vector_matrix_values` which I believe is probably unrelated to my changes.

Reply:
Maybe another instance of numpy/numpy#18500?

Reply:
Tests on Azure passed now. I am resolving this for now, since, the C implementation is now inline with python implementation. Let me know if there are any other concerns regarding this part.